[hipBLASlt]: Add CONTRIBUTING.md to hipBLASlt project (#4596)
ROCm/rocm-libraries: projects/hipblaslt/CONTRIBUTING.md

Reviewer: Some of this is covered in the project README.md not sure if it is of interest to link the two. Also, there is a README.md in tensilelite with config/build info.

**Configure and build from `projects/hipblaslt`** for a focused standalone workflow.

The repo root can build hipblaslt via the superbuild (e.g. `ROCM_LIBS_ENABLE_COMPONENTS`), but that configures additional projects. For a faster dev loop, configure directly in `projects/hipblaslt`:

```bash
cd rocm-libraries/projects/hipblaslt   # or your path to the hipblaslt project

cmake -B build -S . \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_CXX_COMPILER=/opt/rocm/bin/amdclang++ \
  -DCMAKE_C_COMPILER=/opt/rocm/bin/amdclang \
  -DCMAKE_PREFIX_PATH=/opt/rocm \
  -DGPU_TARGETS=gfx90a

cmake --build build --parallel
```

You need a ROCm install (e.g. `/opt/rocm`) and Python 3.8+. For `GPU_TARGETS`, see [GPU targets: one, several, or all](#gpu-targets-one-several-or-all) below.

---

## GPU targets: one, several, or all

**`GPU_TARGETS`** (CMake) controls which AMD GPU architectures are built (Tensile device libraries and any arch-specific code).

- **One GPU:** e.g. `-DGPU_TARGETS=gfx90a`
- **Multiple GPUs:** semicolon-separated, quoted: `-DGPU_TARGETS="gfx90a;gfx942;gfx950"`
- **All supported:** `-DGPU_TARGETS=all` (or leave unset). Takes longer.

> Supported names are in `cmake/tensilelite_supported_architectures.cmake`.

Reviewer: Maybe we could translate the .cmake file to a table here but then we have two sources of truth...

Author: That's a neat idea. Let me think about it, because we could at least have the common ones in here.

### Minimal steps to get tests running (from `projects/hipblaslt`)

```bash
# 0. Python env
python3 -m venv .venv && source .venv/bin/activate
pip install -r tensilelite/requirements.txt

# 1. Configure (use venv Python)
cmake -B build -S . -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_CXX_COMPILER=/opt/rocm/bin/amdclang++ -DCMAKE_PREFIX_PATH=/opt/rocm \
  -DGPU_TARGETS=gfx90a \
  -DPython_EXECUTABLE="$(pwd)/.venv/bin/python" -DPython3_EXECUTABLE="$(pwd)/.venv/bin/python"

# 2. Build device libraries (slow, one-time per config)
cmake --build build --target tensilelite-device-libraries

# 3. Build test binary and generate hipblaslt_gtest.data
cmake --build build --target hipblaslt-test

# 4. Run from build/clients so the data file is found
cd build/clients && ./hipblaslt-test
```

---

## Running the client tests (hipblaslt-test)

### Where does `hipblaslt_gtest.data` come from?

It is **generated at build time**, not checked in. A CMake step runs `clients/tests/hipblaslt_gentest.py` (needs PyYAML) to turn YAML in `clients/tests/data/*.yaml` into `build/clients/hipblaslt_gtest.data`. Building `hipblaslt-test` (or `hipblaslt-test-data`) runs that step.

> **Run from the directory that contains both the test binary and the data file.** The executable looks for `hipblaslt_gtest.data` next to itself (e.g. `build/clients/`):

Reviewer 2: ```suggestion
**Run from the directory that contains both the test binary and the data file.** The executable looks for `hipblaslt_gtest.data` in the same directory (e.g. `build/clients/`):
```

```bash
cd projects/hipblaslt/build/clients
./hipblaslt-test
./hipblaslt-test --gtest_filter=*quick*
```

If you run from elsewhere, the test may not find the data file. For more run options and client details, see the [hipBLASLt clients](https://rocm.docs.amd.com/projects/hipBLASLt/en/latest/conceptual/hipblaslt-clients.html) documentation.

---

## Test data and adding/changing tests

- **YAML data:** Test cases are driven by YAML under `clients/tests/data/` (e.g. `matmul_gtest.yaml`, `hipblaslt_common.yaml`). The build generates `hipblaslt_gtest.data` from these.

> - **Test coverage:** For an overview of what the tests cover and notable gaps, see [docs/TEST_COVERAGE_OVERVIEW.md](docs/TEST_COVERAGE_OVERVIEW.md) when present; otherwise inspect the YAML under `clients/tests/data/` and the test sources under `clients/`.

Reviewer 2: This link doesn't work. It's probably best not to make this a real link if it's not guaranteed to exist, just put the hypothetical link in as quoted text.

Author: Yeah, and if the file isn't there, it's confusing! I just removed that whole clause.
